keps/sig-instrumentation/3466-kubernetes-component-health-slis/README.md

```diff
@@ -0,0 +1,320 @@
+1. type - one of `livez`, `readyz`, `healthz`
+1. name - the known name of the health check. AFAIK, these are all static strings in the Kubernetes codebase, therefore bounded in cardinality
+
+
+## Design Details
+
+When healthz/livez/readyz paths are accessed (not on a timer), they will record whatever they return
+in a gauge metric. Admittedly, this has the downside of staleness though, since the health check 
+data can be as stale as the length of the kubelet scrape interval. However, given our e2e tests 
+configure [apiserver to 1s intervals](https://github.com/kubernetes/kubernetes/blob/master/cluster/gce/manifests/kube-apiserver.manifest#L58), it is reasonable to assume that other cloud-providers 
+likely configure similar small scrape intervals, which means staleness should not realistically 
+be much of an issue. However, in the case that the kubelet gets stuck, one can alert off of the
+counter that we expose; if the counter stops incrementing, then we know that the health endpoint
+is not getting hit and that our gauge data is too stale. It would therefore be prudent to set a
+staleness alert off of the counter.
+
+We considered fetching metric data when the metrics endpoint was hit, but this would introduce 
+extra load against the health endpoint, which we took care to avoid. Alternatively, we considered
+periodic polling of the metrics endpoint such that the metrics would be incremented only during this
+periodic poll, but that change would be larger and would need to be implemented in each component
+for each of their health check endpoints. 
+
+Using a gaugeFunc would also preclude making the metric `stable`, since gaugeFuncs are
+dynamic by nature and therefore cannot be parsed at compile time by the stability framework. Since
+these metrics are intended to be used as component health SLIs, we want them to be able to
+be promoted to `stable` status.
+
+### Test Plan
+
+
+##### Unit tests
+
+- [ ] ensure that healthcheck states are reset for gauges on write
+- [ ] ensure that counters properly retain state of all seen healthchecks
+
+##### Integration tests
+
+- [ ] ensure existence of healthcheck endpoint
+
+##### e2e tests
+
+
+- [ ] ensure existence of healthcheck endpoint
+
+
+### Graduation Criteria
+
+
+## Production Readiness Review Questionnaire
+
+
+### Feature Enablement and Rollback
+
+We can feature gate the api and the registry for this metrics endpoint, such that it is disabled by default
+and toggleable only by feature-gate. Given the lazy nature of the Kubernetes metric framework, this will
+ensure that we do not encounter any memory leaks.
+
+
+###### How can this feature be enabled / disabled in a live cluster?
+
+Via feature gate.
+
+
+###### Does enabling the feature change any default behavior?
+
+Yes, it exposes an endpoint which can then be scraped.
+
+###### Can the feature be disabled once it has been enabled (i.e. can we roll back the enablement)?
+
+Yes, but it will remove the endpoint.
+
```

Comment:
Please fill in all questions from the template:
https://github.com/kubernetes/enhancements/blob/master/keps/NNNN-kep-template/README.md#feature-enablement-and-rollback

Reply:
done